dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/DynamicSpinLoopDetection.java

```diff
@@ -72,167 +70,108 @@ private List<LoopData> computeLoopData(Function func, LoopAnalysis loopAnalysis)
-        loop.getStart().insertAfter(init);
-        for (Event sideEffect : loop.sideEffects) {
+        // Track global side effects
+        for (Event sideEffect : loop.globalSideEffects) {
             final List<Event> updateSideEffect = new ArrayList<>();
             if (sideEffect.cfImpliesExec()) {
-                updateSideEffect.add(EventFactory.newLocal(trackingReg, expressions.makeTrue()));
+                updateSideEffect.add(EventFactory.newLocal(globalSideEffectReg, expressions.makeTrue()));
             } else {
                 updateSideEffect.addAll(List.of(
                         EventFactory.newExecutionStatus(tempReg, sideEffect),
-                        EventFactory.newLocal(trackingReg, expressions.makeOr(trackingReg, expressions.makeNot(tempReg)))
+                        EventFactory.newLocal(globalSideEffectReg, expressions.makeOr(globalSideEffectReg, expressions.makeNot(tempReg)))
                 ));
 
             }
             sideEffect.getPredecessor().insertAfter(updateSideEffect);
         }
 
-        final Event assumeSideEffect = newSpinTerminator(expressions.makeNot(trackingReg), func);
-        assumeSideEffect.copyAllMetadataFrom(loop.getStart());
-        loop.getEnd().getPredecessor().insertAfter(assumeSideEffect);
+        // Check if any local or global side effects occurred. If not, spin!
+        final Register localSideEffectReg = func.newRegister("__localSideEffect#" + loopNum, types.getBooleanType());
+        final Expression hasSideEffect = expressions.makeOr(localSideEffectReg, globalSideEffectReg);
+
+        final Event assignLocalSideEffectReg = EventFactory.newLocal(localSideEffectReg, expressions.makeNEQ(entryLiveStateRegister, liveRegistersVector));
+        final Event assumeSideEffect = newSpinTerminator(expressions.makeNot(hasSideEffect), loop);
+        loop.getEnd().getPredecessor().insertAfter(List.of(
+                assignLocalSideEffectReg,
+                assumeSideEffect
+        ));
 
         // Special case: If the loop is fully side-effect-free, we can set its unrolling bound to 1.
-        if (loop.sideEffects.isEmpty()) {
+        if (loop.isSideEffectFree()) {
             final Event loopBound = EventFactory.Svcomp.newLoopBound(expressions.makeValue(1, types.getArchType()));
             loop.getStart().getPredecessor().insertAfter(loopBound);
         }
     }
 
-    private Event newSpinTerminator(Expression guard, Function func) {
+    private Event newSpinTerminator(Expression guard, LoopData loop) {
+        final Function func = loop.getStart().getFunction();
         final Event terminator = func instanceof Thread thread ?
                 EventFactory.newJump(guard, (Label) thread.getExit())
                 : EventFactory.newAbortIf(guard);
-        terminator.addTags(Tag.SPINLOOP, Tag.NONTERMINATION, Tag.NOOPT);
+        terminator.addTags(Tag.SPINLOOP, Tag.NONTERMINATION);
```

Comment:
Why did you remove the `Tag.NOOPT` tag?

Reply:
It's a good question and I don't know the exact answer to this. The first version of Spinloop detection didn't have it and it is not needed in principle. I only vaguely remember that I added the NOOPT a while ago to work around some issue, maybe related to `SyntacticContextAnalysis`?

Reply:
I tracked the history of this pass. The pass has been renamed / moved several times, but I think [this](https://github.com/hernanponcedeleon/Dat3M/commit/05c5d18626227f45765a0dc5bbcb1faa8b742935#diff-43ac8b13a92d3b7c3af2ac0ef91c6582b30c172ed8a157708f3123eb7b3059dd) is the commit where the tag was added. It is not really clear if the tag was needed specifically for spinloops.